packages/tpetra/core/src/Tpetra_Import_Util2.hpp

```diff
@@ -1237,28 +1250,263 @@ lowCommunicationMakeColMapAndReindexKokkos (const Teuchos::ArrayView<const size_
+      auto localColMap = colMap->getLocalMap();
+      //Kokkos::parallel_for(Kokkos::RangePolicy<execution_space>(0, colind_GID.size()), KOKKOS_LAMBDA(const int i) {
+      Kokkos::parallel_for(Kokkos::RangePolicy<execution_space>(0, colind_GID_view.size()), KOKKOS_LAMBDA(const int i) {
+        colind_LID_view[i] = localColMap.getLocalElement(colind_GID_view[i]);
+      });
+      //Kokkos::deep_copy(execution_space(), colind_LID_host, colind_LID_view);
+      return;
+    }
   }
-  else lowCommunicationMakeColMapAndReindexKokkos<LocalOrdinal, GlobalOrdinal, Node> (rowptr, colind_LID, colind_GID, domainMapRCP, owningPIDs, remotePIDs, colMap);
-}
 
+  // Now build the array containing column GIDs
+  // Build back end, containing remote GIDs, first
+  const LO numMyCols = NumLocalColGIDs + NumRemoteColGIDs;
+  Kokkos::View<GO*, DT> ColIndices_view("ColIndices", numMyCols);
+  auto ColIndices_host = Kokkos::create_mirror_view(ColIndices_view);
+  
+  // We don't need to load the backend of ColIndices or sort if there are no remote GIDs
+  if(NumRemoteColGIDs > 0) {
+    if(NumLocalColGIDs != static_cast<size_t> (numMyCols)) {
+      Kokkos::parallel_for(Kokkos::RangePolicy<execution_space>(0, NumRemoteColGIDs), KOKKOS_LAMBDA(const int i) {
+        ColIndices_view[NumLocalColGIDs+i] = RemoteGIDList_view[i];
+      });
+    }  
+  
+    // Find the largest PID for bin sorting purposes
+    int PID_max = 0;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<execution_space>(0, PIDList_host.size()), KOKKOS_LAMBDA(const int i, int& max) {
+      if(max < PIDList_view[i]) max = PIDList_view[i];
+    }, Kokkos::Max<int>(PID_max));
+  
+    using KeyViewTypePID = decltype(PIDList_view);
+    using BinSortOpPID = Kokkos::BinOp1D<KeyViewTypePID>;
+  
+    // Make a subview of ColIndices for remote GID sorting
+    auto ColIndices_subview = Kokkos::subview(ColIndices_view, Kokkos::make_pair(NumLocalColGIDs, ColIndices_view.size()));
+  
+    // Make binOp with bins = PID_max + 1, min = 0, max = PID_max
+    BinSortOpPID binOp2(PID_max+1, 0, PID_max);
+    
+    // Sort External column indices so that all columns coming from a
+    // given remote processor are contiguous.  This is a sort with one
+    // auxilary array: RemoteColIndices
+    Kokkos::BinSort<KeyViewTypePID, BinSortOpPID> bin_sort2(PIDList_view, 0, PIDList_view.size(), binOp2, false);
+    bin_sort2.create_permute_vector(exec);
+    bin_sort2.sort(exec, PIDList_view);
+    bin_sort2.sort(exec, ColIndices_subview);
+  
+    // Deep copy back from device to host
+    Kokkos::deep_copy(execution_space(), PIDList_host, PIDList_view);
+  
+    // Stash the RemotePIDs. Once remotePIDs is changed to become a Kokkos view, we can remove this and copy directly.
+    // Note: If Teuchos::Array had a shrink_to_fit like std::vector,
+    // we'd call it here.
+    
+    Teuchos::Array<int> PIDList(NumRemoteColGIDs);
+    for(LO i = 0; i < NumRemoteColGIDs; ++i) {
+      PIDList[i] = PIDList_host[i];
```

Comment:
Need a fence on ``execution_space()`` before accessing ``PIDLIST_host`` - the deep_copy above could still be running at this point

Reply:
BTW, you might as well copy directly from PIDList_view to PIDList using an unmanaged view (still need the fence, but avoid this host-host copy)
```
Teuchos::Array<int> PIDList(NumRemoteColGIDs);
Kokkos::View<int*, Kokkos::HostSpace> PIDList_host(PIDList.data(), PIDList.size());
Kokkos::deep_copy(execution_space(), PIDList_host, PIDList_view);
```

Reply:
done